### DISCUSSION TITLE
Get git diff of files in actions

### DISCUSSION THREAD

**Discussion by thatrandomperson5** [community/community#30950] [Actions] [answered]
I need to get all of the diff of files from a pr to an env var. How would I do this? I have the pr event.

Thanks, thatrnpr5

**Comment by airtower-luna** [comment, marked as answer]
Look at the `pull_request` object inside the event (`github.event.pull_request`), it's [described in the REST API docs](https://docs.github.com/en/rest/pulls/pulls#get-a-pull-request). Most importantly, it has `base` and `head` elements describing the base and head branches of the PR. Get the `sha` element of both and compare as usual.

The trick is to make `actions/checkout` fetch those commits first. Because they should be ancestors of the temporary merge commit `fetch-depth: 2` should be enough.

**Reply by thatrandomperson5** [reply]
What is sha? I lost you after this part: The trick is to make actions/checkout fetch those commits first.

**Reply by airtower-luna** [reply]
> What is sha?

The commit ID, which is an SHA-1 hash of the commit object.

> I lost you after this part: The trick is to make actions/checkout fetch those commits first.

By default `actions/checkout` (which you usually use in a workflow to get the repository) gets only the commit that triggered the workflow run. For a `pull_request` event that's a temporary merge commit. Basically what would be the result of merging.

I strongly recommend you look at the "[Git Book](https://git-scm.com/book/en/v2)" site to learn about concepts like commit IDs and merging. Chapters 2 and 3 cover the most important things.

**Reply by thatrandomperson5** [reply]
I found out sha afterward, i know id's and merging. Where should i add fetch depth 2? And do i need a token to use the github api if im just looking up stuff.

**Reply by airtower-luna** [reply]
> Where should i add fetch depth 2?

To the `actions/checkout` step in your workflow, like:

```yaml
- uses: actions/checkout@v3
  with:
    fetch-depth: 2
```

> And do i need a token to use the github api if im just looking up stuff.

The default `GITHUB_TOKEN` provided by the workflow should be enough. But you shouldn't need to use the API beyond the usual checkout (just with slightly more depth) anyway, all the other data is already in the event payload. I just pointed at the API docs because the structure of the `pull_request` object is described there.

**Reply by thatrandomperson5** [reply]
So something like 
```yaml

env:
     COMMITS_B: ${{ github.event.pull_request.base.sha }}
     COMMITS_H: ${{ github.event.pull_request.head.sha }}

```
?
but that gives me the id, not the content. How would I get the content?

**Reply by airtower-luna** [reply]
That's why I mentioned you should use `git diff`, or whatever other command you need to create the kind of diff you want. You need the commit IDs to tell `git diff` what to compare, and you need the increased `fetch-depth` so those commits are available to compare.

With those environment variables you could for example do `git diff ${COMMITS_B}..${COMMITS_H}` in a `run` step, and direct the output wherever you like. Using an environment variable to hold a diff seems questionable, but if you really want to use the [multiline form of writing to `$GITHUB_ENV`](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#multiline-strings).

**Reply by thatrandomperson5** [reply]
When I save git diff I get `diff --git a/public/refs.toml b/public/refs.toml ` instead of the result.
Command ran: `echo "::set-output name=diff::$(git diff ${{ github.event.pull_request.base.sha }}..${{ github.event.pull_request.head.sha }} -- ./public/refs.toml)" `

**Reply by airtower-luna** [reply]
The `set-output` command breaks with multiline data. You need to URL-encode the line breaks if you want to use a step output.

I don't know the rest of your workflow, of course, but it's probably much easier if you redirect the output of `git diff` to a file and do whatever further processing you need on that file. :sweat_smile: 

```
git diff ${{ github.event.pull_request.base.sha }}..${{ github.event.pull_request.head.sha }} -- ./public/refs.toml >refs.diff
```

Then you have the diff in `refs.diff` and can push it around easily.

**Reply by thatrandomperson5** [reply]
One last thing, I can’t seem to figure out how to remove the extra stuff from git diff.
I get:
```diff
diff --git a/public/refs.toml b/public/refs.toml
index 3cfb11e..ba17439 100644
--- a/public/refs.toml
+++ b/public/refs.toml
@@ -2,3 +2,4 @@
 #Format: name = "url"
 [github]
 thatrnpr = {val = "https://github.com/thatrandomperson5", author = "thatrandomperson5"}
+lol = {val = "https://github.com/l/o/l", author = "thatrandomperson5"}
```
I want:
```diff
 #Format: name = "url"
 [github]
 thatrnpr = {val = "https://github.com/thatrandomperson5", author = "thatrandomperson5"}
+lol = {val = "https://github.com/l/o/l", author = "thatrandomperson5"}

```
I also wan’t all lines in one batch. Not multiple with the @@ thing.
What arguments should I use/how should I go about doing this?

**Reply by airtower-luna** [reply]
The "@@ thing" is critical to the diff working. What you can do is [increase the number of context lines](https://git-scm.com/docs/git-diff#Documentation/git-diff.txt--Ultngt) to effectively make it one batch. The number of lines in the file works pretty well, `wc` lets you get that easily:

```
git diff -U"$(wc -l FILE | cut -d' ' -f1)" FROM..TO -- FILE
```

As long as you're only diff'ing one file, you could just cut off the first 4 lines if you don't need them. Whether it makes sense to do depends on your context. If you want to show a diff to someone I'd strongly recommend just leaving it as is.
